src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -15,23 +17,53 @@ private static void HandleStartObject(JsonSerializerOptions options, ref ReadSta
                 return;
             }
 
-            if (state.Current.IsDictionary())
-            {
-                // Fall through and treat as a return value.
-            }
-            else if (state.Current.IsEnumerable() || state.Current.IsPropertyEnumerable() || state.Current.IsPropertyADictionary())
+            if (state.Current.IsProcessingEnumerable())
             {
-                // An array of objects either on the current property or on a list
                 Type objType = state.Current.GetElementType();
                 state.Push();
-                state.Current.JsonClassInfo = options.GetOrAddClass(objType);
+                state.Current.Initialize(objType, options);
             }
             else if (state.Current.JsonPropertyInfo != null)
             {
-                // Nested object
-                Type objType = state.Current.JsonPropertyInfo.RuntimePropertyType;
-                state.Push();
-                state.Current.JsonClassInfo = options.GetOrAddClass(objType);
+                if (state.Current.IsDictionary())
+                {
+                    // Verify that the Dictionary can be deserialized.
+                    if (state.Current.JsonClassInfo.Type.GetGenericArguments()[0].UnderlyingSystemType != typeof(string))
+                    {
+                        ThrowHelper.ThrowJsonReaderException_DeserializeUnableToConvertValue(state.Current.JsonClassInfo.Type, reader, state);
+                    }
+
+                    ClassType classType = state.Current.JsonClassInfo.ElementClassInfo.ClassType;
+
+                    if (state.Current.ReturnValue == null)
+                    {
+                        // The Dictionary created below will be returned to corresponding Parse() etc method.
+                        // Ensure any nested array creates a new frame.
+                        state.Current.EnumerableCreated = true;
+                    }
+                    else if (classType == ClassType.Object || classType == ClassType.Dictionary)
+                    {
+                        // A nested object or dictionary.
+                        JsonClassInfo classInfoTemp = state.Current.JsonClassInfo;
+                        state.Push();
+                        state.Current.JsonClassInfo = classInfoTemp.ElementClassInfo;
+                        state.Current.InitializeJsonPropertyInfo();
+                    }
+                    else
+                    {
+                        // We shouldn't get here unless we add to the ClassType enum.
+                        Debug.Assert(false);
+                        ThrowHelper.ThrowJsonReaderException_DeserializeUnableToConvertValue(
```

Comment:
If we are expected not to get here, why are we throwing? Should this instead just be a message in the `Debug.Assert`?

Reply:
A bit of defensive programming here during development. I can remove it for now. This code will need to be modified if we want to support dictionary properties that do not contain a setter (I am creating an issue for that).